### PR TITLE
LogicSig support

### PR DESCRIPTION
**Pull request by pzbitskiy**
Implemented functionality for LogicSig creation (both delegated and non-delegated)
Added usage example to README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -698,4 +698,77 @@ func submitGroup() {
...
+	"github.com/algorand/go-algorand-sdk/mnemonic"
+	"github.com/algorand/go-algorand-sdk/transaction"
+	"github.com/algorand/go-algorand-sdk/types"
+)
+
+// CHANGE ME
+const algodAddress = "http://localhost:8080"
+const algodToken = "6218386c0d964e371f34bbff4adf543dab14a7d9720c11c6f11970774d4575de"
+
+func doLogSig() {
```
Comment (pzbitskiy) [inline README.md]: should be `func main()`

Comment (pzbitskiy) [reply README.md]: done

File: crypto/crypto.go [hunk 2]
```diff
@@ -309,3 +376,164 @@ func ComputeGroupID(txgroup []types.Transaction) (gid types.Digest, err error) {
...
+
+	hasSig := false
+	hasMsig := false
+
+	if lsig.Sig != (types.Signature{}) {
+		hasSig = true
+	}
+	if !lsig.Msig.Blank() {
+		hasMsig = true
+	}
```
Comment (rotemh) [inline crypto/crypto.go]: Consider simplifying to 
```go
hasSig := lsig.Sig != (types.Signature{})
hasMsig := !lsig.Msig.Blank()
```

**Comment by rotemh** [comment]
A few comments but looks good

File: crypto/crypto.go [hunk 3]
```diff
@@ -147,6 +151,40 @@ func SignBid(sk ed25519.PrivateKey, bid types.Bid) (signedBid []byte, err error)
 
 /* Multisig Support */
 
+type signer func() (signature types.Signature, err error)
+
+// Service function to make a single signature in Multisig
+func multisigSingle(sk ed25519.PrivateKey, pk MultisigAccount, customSigner signer) (msig types.MultisigSig, myIndex int, err error) {
+	// check that sk.pk exists in the pk list
+	myIndex = len(pk.Pks)
```
[the remark this answers is not in the document]
Comment (pzbitskiy) [reply crypto/crypto.go]: done

File: crypto/account.go [hunk 4]
```diff
@@ -115,3 +115,31 @@ func (ma MultisigAccount) Validate() (err error) {
...
+
+// LogicSigAddress returns contract (escrow) address
+func LogicSigAddress(lsig types.LogicSig) types.Address {
+	toBeSigned := programToSign(lsig.Logic)
+	checksum := sha512.Sum512_256(toBeSigned)
+
+	var addr types.Address
+	n := copy(addr[:], checksum[:])
+	if n != ed25519.PublicKeySize {
+		panic("generated public key is the wrong size")
```
[the remark this answers is not in the document]
Comment (pzbitskiy) [reply crypto/account.go]: done

File: crypto/crypto.go [hunk 5]
```diff
@@ -289,6 +311,51 @@ func AppendMultisigTransaction(sk ed25519.PrivateKey, pk MultisigAccount, preStx
...
+	for _, subsigi := range msig.Subsigs {
+		if (subsigi.Sig != types.Signature{}) {
+			if !ed25519.Verify(subsigi.Key, message, subsigi.Sig[:]) {
+				return false
+			}
+			verifiedCount++
+		}
+	}
+
+	if verifiedCount < int(msig.Threshold) {
```
[the remark this answers is not in the document]
Comment (pzbitskiy) [reply crypto/crypto.go]: done

File: crypto/crypto.go [hunk 6]
```diff
@@ -309,3 +376,164 @@ func ComputeGroupID(txgroup []types.Transaction) (gid types.Digest, err error) {
...
+	toBeSigned := programToSign(lsig.Logic)
+	// logic sig, compare hashes
+	if !hasSig && !hasMsig {
+		result = types.Digest(sha512.Sum512_256(toBeSigned)) == types.Digest(sender)
+		return
+	}
+
+	if hasSig {
+		result = ed25519.Verify(sender[:], toBeSigned, lsig.Sig[:])
+	} else {
```
[the remark this answers is not in the document]
Comment (pzbitskiy) [reply crypto/crypto.go]: done

File: crypto/crypto.go (hunk 2 above)
Comment (pzbitskiy) [reply crypto/crypto.go]: done

File: crypto/crypto.go [hunk 7]
```diff
@@ -147,6 +151,40 @@ func SignBid(sk ed25519.PrivateKey, bid types.Bid) (signedBid []byte, err error)
...
+	// check that sk.pk exists in the pk list
+	myIndex = len(pk.Pks)
+	myPk := sk.Public().(ed25519.PublicKey)
+	for i := 0; i < len(pk.Pks); i++ {
+		if bytes.Equal(myPk, pk.Pks[i]) {
+			myIndex = i
+		}
+	}
+	if myIndex == len(pk.Pks) {
+		err = errMsigInvalidSecretKey
```
[the remark this answers is not in the document]
Comment (pzbitskiy) [reply crypto/crypto.go]: done

**Comment by pzbitskiy** [comment]
@rotemh fixed remarks and rebased

**Pull request merged**